### PR TITLE
init bulletml and rrootage

### PR DESCRIPTION
**Pull request by fgaz**
###### Motivation for this change

I'm packaging some of the awesome games by Kenta Cho. This is the first.

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

**Comment by Mic92** [comment]
@GrahamcOfBorg build bulletml rrootage

File: pkgs/development/libraries/bulletml/default.nix [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+  postUnpack = "mv d_cpp bulletml/";
+
+  patches = [
+    (debianPatch "fixes" "0cnr968n0h50fjmjijx7idsa2pg2pv5cwy6nvfbkx9z8w2zf0mkl")
+    (debianPatch "bulletml_d" "03d1dgln3gkiw019pxn3gwgjkmvzisq8kp3n6fpn38yfwh4fp4hv")
+    (debianPatch "d_cpp" "04g9c7c89w7cgrxw75mcbdhzxqmz1716li49mhl98znakchrlb9h")
+    (debianPatch "warnings" "18px79x4drvm6dy6w6js53nzlyvha7qaxhz5a99b97pyk3qc7i9g")
+    (debianPatch "makefile" "0z6yxanxmarx0s08gh12pk2wfqjk8g797wmfcqczdv1i6xc7nqzp")
+    (debianPatch "includes" "1n11j5695hs9pspslf748w2cq5d78s6bwhyl476wp6gcq6jw20bw")
+    (debianPatch "get-rid-of-boost" "1wc8jxlxmdx53mhgx55wzysxxbbrdmq6dpfbb8rrrbkdz5b59hjm")
```
Comment (Mic92) [inline pkgs/development/libraries/bulletml/default.nix]: Have those patches been reviewed if they apply to Nixos as well?

File: pkgs/games/rrootage/default.nix [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+  };
+
+in stdenv.mkDerivation {
+  name = "rrootage-${version}";
+  src = fetchurl {
+    url = "http://downloads.sourceforge.net/rrootage/rRootage-${version}.tar.gz";
+    sha256 = "01zzg4ih3kmbhsn1p9zr7g8srv1d2dhrp8cdd86y9qq233idnkln";
+  };
+
+  patches = [
```
Comment (Mic92) [inline pkgs/games/rrootage/default.nix]: Same here.

File: pkgs/development/libraries/bulletml/default.nix (hunk 1 above)
Comment (fgaz) [reply pkgs/development/libraries/bulletml/default.nix]: Not individually. I know that without them it does not compile and with all of them it works fine. I'll review them one by one and remove the unnecessary ones.

File: pkgs/games/rrootage/default.nix (hunk 2 above)
Comment (fgaz) [reply pkgs/games/rrootage/default.nix]: I'm checking right now, but I'm unsure about some.

For example removing [this patch](https://sources.debian.org/src/rrootage/0.23a-12/debian/patches/10_deg_out_of_range.patch/) seems to have no effect on the runtime behavior of rrootage, but looking at the description it makes sense to include it anyway.

What do you think I should do?

Comment (fgaz) [reply pkgs/games/rrootage/default.nix]: Or the "high res by default" patch or the "windowed mode" patch, which are both _really_ useful but not strictly necessary

**Review by matthewbauer — CHANGES_REQUESTED** [review]
See @GrahamcOfBorg build log

File: pkgs/games/rrootage/default.nix (hunk 2 above)
Comment (fgaz) [reply pkgs/games/rrootage/default.nix]: @Mic92 ping?

Maybe I should just fork that repository and apply those patches...

**Comment by Mic92** [comment]
@GrahamcOfBorg build bulletml rrootage

File: pkgs/development/libraries/bulletml/default.nix [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  ];
+
+  makeFlags = [ "-C src" ];
```
Comment (Mic92) [inline pkgs/development/libraries/bulletml/default.nix]: Adding `"CC=c++"`hopefully fixes the macOS build.

**Comment by Mic92** [comment]
@GrahamcOfBorg build bulletml rrootage

File: pkgs/development/libraries/bulletml/default.nix [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    (debianPatch "d_cpp" "04g9c7c89w7cgrxw75mcbdhzxqmz1716li49mhl98znakchrlb9h")
+    (debianPatch "warnings" "18px79x4drvm6dy6w6js53nzlyvha7qaxhz5a99b97pyk3qc7i9g")
+    (debianPatch "makefile" "0z6yxanxmarx0s08gh12pk2wfqjk8g797wmfcqczdv1i6xc7nqzp")
+    (debianPatch "includes" "1n11j5695hs9pspslf748w2cq5d78s6bwhyl476wp6gcq6jw20bw")
+    (debianPatch "get-rid-of-boost" "1wc8jxlxmdx53mhgx55wzysxxbbrdmq6dpfbb8rrrbkdz5b59hjm")
+  ];
+
+  makeFlags = [
+    "-C src"
+    "CC=c++" # Needed for darwin
```
Comment (Mic92) [inline pkgs/development/libraries/bulletml/default.nix]: My bad `CXX=c++` and `CC=cc` hopefully produce better results.

**Comment by Mic92** [comment]
@GrahamcOfBorg build bulletml rrootage

**Comment by Mic92** [comment]
@GrahamcOfBorg build bulletml rrootage

**Comment by Mic92** [comment]
@GrahamcOfBorg build bulletml rrootage

**Comment by Mic92** [comment]
Looks good to me now. If you could squash my trial and errors back into the bulletml commit this pull request should be good to go.

**Comment by Mic92** [comment]
Ah no. now we see more c++ legacy stuff: https://github.com/cuberite/cuberite/issues/424

**Comment by Mic92** [comment]
environment or not is the same thing in make. Adding the `-std=c++11` flag might help to solve include problem.

**Comment by fgaz** [comment]
I just discovered that it was the remove-boost debian patch that introduced the tr1/memory include. I removed that patch and installed the boost files too.

Could you try to rebuild now, @Mic92 ?

**Comment by Mic92** [comment]
@GrahamcOfBorg build bulletml rrootage

**Comment by fgaz** [comment]
The x86_64-linux failure is spurious.

darwin says `ld: unknown option: -soname`. From what I found it seems to require -install_name instead of -soname

**Comment by fgaz** [comment]
@Mic92 last try, could you trigger a rebuild?

If that fails can we merge this as is? Maybe with platforms=linux and a comment referencing this discussion.

Without a darwin machine to edit-test-iterate this on, fixing it is a long and painful process

**Comment by fgaz** [comment]
I'm not even sure if that builtins.currentSystem comparison does what I think it does

**Comment by Mic92** [comment]
@GrahamcOfBorg build bulletml rrootage

**Comment by Mic92** [comment]
Then please disable macOS as platform.

**Comment by fgaz** [comment]
@Mic92 done & removed the commits relevant to darwin.

For anyone that comes across this and wants to fix the darwin build here's what was tried:

```diff
+  postPatch = if builtins.currentSystem == stdenv.lib.platforms.darwin
+              then "substituteInPlace Makefile --replace soname install_name"
+              else null;
 
   makeFlags = [
     "-C src"
+    # Needed for darwin
+    "CC=cc"
+    "CXX=c++"
+    "LD=c++"
   ];
```

But the currentSystem comparison is always false for some reason

**Pull request merged**